fxevent/slog.go

```diff
@@ -0,0 +1,277 @@
+				slogStrings("stacktrace", e.StackTrace),
+				slogStrings("moduletrace", e.ModuleTrace),
+				slogMaybeModuleField(e.ModuleName),
+				slog.String("type", rtype),
+			)
+		}
+		if e.Err != nil {
+			l.logError("error encountered while replacing",
+				slogStrings("stacktrace", e.StackTrace),
+				slogStrings("moduletrace", e.ModuleTrace),
+				slogMaybeModuleField(e.ModuleName),
+				slogErr(e.Err))
+		}
+	case *Decorated:
+		for _, rtype := range e.OutputTypeNames {
+			l.logEvent("decorated",
+				slog.String("decorator", e.DecoratorName),
+				slogStrings("stacktrace", e.StackTrace),
+				slogStrings("moduletrace", e.ModuleTrace),
+				slogMaybeModuleField(e.ModuleName),
+				slog.String("type", rtype),
+			)
+		}
+		if e.Err != nil {
+			l.logError("error encountered while applying options",
+				slogStrings("stacktrace", e.StackTrace),
+				slogStrings("moduletrace", e.ModuleTrace),
+				slogMaybeModuleField(e.ModuleName),
+				slogErr(e.Err))
+		}
+	case *Run:
+		if e.Err != nil {
+			l.logError("error returned",
+				slog.String("name", e.Name),
+				slog.String("kind", e.Kind),
+				slogMaybeModuleField(e.ModuleName),
+				slogErr(e.Err),
+			)
+		} else {
+			l.logEvent("run",
+				slog.String("name", e.Name),
+				slog.String("kind", e.Kind),
+				slogMaybeModuleField(e.ModuleName),
+			)
+		}
+	case *Invoking:
+		// Do not log stack as it will make logs hard to read.
+		l.logEvent("invoking",
+			slog.String("function", e.FunctionName),
+			slogMaybeModuleField(e.ModuleName),
+		)
+	case *Invoked:
+		if e.Err != nil {
+			l.logError("invoke failed",
+				slogErr(e.Err),
+				slog.String("stack", e.Trace),
+				slog.String("function", e.FunctionName),
+				slogMaybeModuleField(e.ModuleName),
+			)
+		}
+	case *Stopping:
+		l.logEvent("received signal",
+			slog.String("signal", strings.ToUpper(e.Signal.String())))
+	case *Stopped:
+		if e.Err != nil {
+			l.logError("stop failed", slogErr(e.Err))
+		}
+	case *RollingBack:
+		l.logError("start failed, rolling back", slogErr(e.StartErr))
+	case *RolledBack:
+		if e.Err != nil {
+			l.logError("rollback failed", slogErr(e.Err))
+		}
+	case *Started:
+		if e.Err != nil {
+			l.logError("start failed", slogErr(e.Err))
+		} else {
+			l.logEvent("started")
+		}
+	case *LoggerInitialized:
+		if e.Err != nil {
+			l.logError("custom logger initialization failed", slogErr(e.Err))
+		} else {
+			l.logEvent("initialized custom fxevent.Logger", slog.String("function", e.ConstructorName))
+		}
+	}
+}
+
+type slogFieldSkip struct{}
+
+func slogMaybeModuleField(name string) slog.Attr {
+	if len(name) == 0 {
+		return slog.Any("module", slogFieldSkip{})
+	}
+	return slog.String("module", name)
+}
+
+func slogMaybeBool(name string, b bool) slog.Attr {
+	if !b {
+		return slog.Any(name, slogFieldSkip{})
+	}
+	return slog.Bool(name, true)
+}
+
+func slogErr(err error) slog.Attr {
+	return slog.String("error", err.Error())
+}
+
+func slogStrings(key string, str []string) slog.Attr {
+	var attrs []any
```

Comment:
can we pre-allocate here?
```suggestion
	attrs := make([]slog.Attr, 0, len(str))
```

Reply:
yup! resolved